src/core/server/plugins/discovery/plugin_manifest_parser.ts

```diff
@@ -38,6 +38,27 @@ const MANIFEST_FILE_NAME = 'kibana.json';
  */
 const ALWAYS_COMPATIBLE_VERSION = 'kibana';
 
+/**
+ * Names of the known manifest fields.
+ */
+const KNOWN_MANIFEST_FIELDS = (() => {
```

Comment:
__note:__ couldn't find a better solution, have to hear any suggestions :/ Unfortunately something like `const KNOWN_MANIFEST_FIELDS: Array<keyof PluginManifest> = ['server', .....]` doesn't work since it won't complain if we forget to include a property.

Reply:
How would you feel about using `@kbn/config-schema` instead of trying to verify the extraction of the keys from an interface? Write the schema and then base the interface on the schema.

Reply:
So it'd look like this:
```ts
const PluginManifestSchema = schema.object(
  {
    id: schema.string({
      validate(id) {
        if (id.includes('.')) {
          return 'value must not include `.` characters.';
        }
      },
    }),
    configPath: schema.oneOf(
      [
        schema.string({
          validate(path) {
            if (path.includes('.')) {
              return 'value must not include `.` characters.';
            }
          },
        }),
        schema.arrayOf(
          schema.string({
            validate(path) {
              if (path.includes('.')) {
                return 'value must not include `.` characters.';
              }
            },
          })
        ),
      ],
      { defaultValue: schema.siblingRef('id') }
    ),
    version: schema.string(),
    kibanaVersion: schema.string({
      defaultValue: schema.siblingRef('version'),
    }),
    requiredPlugins: schema.arrayOf(schema.string(), { defaultValue: [] }),
    optionalPlugins: schema.arrayOf(schema.string(), { defaultValue: [] }),
    server: schema.boolean({ defaultValue: false }),
    ui: schema.boolean({ defaultValue: false }),
  },
  {
    validate(manifest, { packageVersion }) {
      if (!isVersionCompatible(manifest.kibanaVersion, packageVersion)) {
        return `Plugin "${manifest.id}" is only compatible with Kibana version "${
          manifest.kibanaVersion
        }", but used Kibana version is "${packageVersion}".`;
      }

      if (!manifest.server && !manifest.ui) {
        return `Both "server" and "ui" are missing or set to "false" in plugin manifest for "${
          manifest.id
        }", but at least one of these must be set to "true".`;
      }
    },
  }
);
```
Here is comparison of errors we get before and after using `@kbn-config-schema`:
```
"Plugin manifest must contain a JSON encoded object. (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"
"[manifest]: expected a plain object value, but found [null] instead. (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"

"Plugin manifest must contain an \"id\" property. (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"
"[manifest.id]: expected value of type [string] but got [undefined] (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"

"Plugin \"id\" must not include `.` characters. (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"
"[manifest.id]: value must not include `.` characters. (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"

"Plugin manifest for \"some-id\" must contain a \"version\" property. (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"
"[manifest.version]: expected value of type [string] but got [undefined] (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"

"Plugin \"some-id\" is only compatible with Kibana version \"6.4.2\", but used Kibana version is \"7.0.0-alpha1\". (incompatible-version, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"
"[manifest]: Plugin \"some-id\" is only compatible with Kibana version \"6.4.2\", but used Kibana version is \"7.0.0-alpha1\". (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"

"Plugin \"some-id\" is only compatible with Kibana version \"non-sem-ver\", but used Kibana version is \"7.0.0-alpha1\". (incompatible-version, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"
"[manifest]: Plugin \"some-id\" is only compatible with Kibana version \"non-sem-ver\", but used Kibana version is \"7.0.0-alpha1\". (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"

"The \"configPath\" in plugin manifest for \"some-id\" should either be a string or an array of strings. (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"
"[manifest.configPath]: types that failed validation:
    - [manifest.configPath.0]: expected value of type [string] but got [number]
    - [manifest.configPath.1]: expected value of type [array] but got [number] (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"

The \"configPath\" in plugin manifest for \"some-id\" should either be a string or an array of strings. (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"
"[manifest.configPath]: types that failed validation:
    - [manifest.configPath.0]: expected value of type [string] but got [Array]
    - [manifest.configPath.1.1]: expected value of type [string] but got [number] (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"

"Plugin \"some-id\" is only compatible with Kibana version \"7.0.1\", but used Kibana version is \"7.0.0-alpha1\". (incompatible-version, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"
"[manifest]: Plugin \"some-id\" is only compatible with Kibana version \"7.0.1\", but used Kibana version is \"7.0.0-alpha1\". (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"

"Both \"server\" and \"ui\" are missing or set to \"false\" in plugin manifest for \"some-id\", but at least one of these must be set to \"true\". (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"
"[manifest]: Both \"server\" and \"ui\" are missing or set to \"false\" in plugin manifest for \"some-id\", but at least one of these must be set to \"true\". (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"

"Manifest for plugin \"some-id\" contains the following unrecognized properties: unknownOne,unknownTwo. (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"
"[manifest.unknownOne]: definition for this key is missing (invalid-manifest, /projects/elastic/master/kibana/path/existent-dir/kibana.json)"
```

The issues we have with this are:
* Subjectively error messages are more detailed, but look a bit too generic. `{manifest}]` part from `[{manifest}]` can be changed to whatever we want.
* We can't really distinguish different types of errors, so all errors will have `invalid-maniset` type (currently we have two types `invalid-manifest` and `incompatible-version`). We can probably move version check out of `schema` though, so it's not a deal breaker.
* This will require 2 changes in `@kbn/config-schema`: add support for `allowEmpty` option for `schema.string` and pass context to custom validate function. Both seem reasonable to me even without this use case.

What do you think @spalger ?

Reply:
I agree that the errors as slightly less ideal, but I also think they're pretty good/understandable. If you think it feels fine when actually using it I'd like to see the PR. 

> We can't really distinguish different types of errors

There are error codes internally right? For the messages? Why wouldn't those be attached to the errors as properties?

Reply:
> There are error codes internally right? For the messages? Why wouldn't those be attached to the errors as properties?

Internally we have codes like `any.custom` or `string.min` coming from Joi that we drop and convert errors to our own `SchemaTypeError`. Theoretically we can expose `path` to know what field failed and add some code to know what type of validation failed and then use that info to figure out what and why exactly failed. There is also one special custom validation (via `validate` function) that currently just returns error message, it can fail for various reasons and if we want to distinguish these reasons consumers will have to either provide some code in addition to message or use message as the code itself.

I'll probably experiment a little bit more and create PR if I feel it's good enough :)